xwiki-commons-core/xwiki-commons-environment/xwiki-commons-environment-servlet/src/main/java/org/xwiki/environment/internal/ServletEnvironment.java

```diff
@@ -54,6 +62,41 @@ public class ServletEnvironment extends AbstractEnvironment
      */
     private javax.servlet.ServletContext javaxServletContext;
 
+    @Inject
+    private ComponentManager componentManager;
+
+    @Inject
+    private CacheControl cacheControl;
+
+    private Cache<Optional<URL>> resourceURLCache;
+
+    private final AtomicBoolean cacheInitializing = new AtomicBoolean();
+
+    private Cache<Optional<URL>> getResourceURLCache()
```

Comment:
In case you did not have it in mind, an alternative to this complex ordering game would be to put the cache right in a proxy of the ServletContext.